src/lazyfree.c

```diff
@@ -162,13 +161,15 @@ int dbAsyncDelete(redisDb *db, robj *key) {
             bioCreateLazyFreeJob(lazyfreeFreeObject,1, val);
             dictSetVal(db->dict,de,NULL);
         }
-    }
 
-    /* Release the key-val pair, or just the key if we set the val
-     * field to NULL in order to lazy free it later. */
-    if (de) {
+        /* Release the key-val pair, or just the key if we set the val
+        * field to NULL in order to lazy free it later. */
         if (server.cluster_enabled) slotToKeyDelEntry(de);
-        dictFreeUnlinkedEntry(db->dict,de);
+        dictFreePlinkEntry(db->dict,de,plink,table);
+
+        /* Deleting an entry from the expires dict will not free the sds of
+        * the key, because it is shared with the main dictionary. */
+        if (dictSize(db->expires) > 0) dictDelete(db->expires,key->ptr);
```

Comment:
I think you have to delete from the expires dictionary before releasing the main dict entry, since they share the key name pointer, and only the main dict does the freeing part.